### DISCUSSION TITLE
Livewire\Exceptions\ComponentNotFoundException - Unable to find component: [Livewire.Tasks-Board]

### DISCUSSION THREAD

**Discussion by hussain-attari** [livewire/livewire#4136] [1. Help] [answered]
I created the component using the following command

`php artisan make:livewire TasksBoard`

This is the Livewire Tasks component

```
<?php

namespace App\Http\Livewire;

use Livewire\Component;

class TasksBoard extends Component
{
    public $tasks;

    public function mount()
    {
        $this->tasks = auth()->user()->statuses()->with('tasks')->get();
    }

    public function render()
    {
        return view('livewire.tasks-board');
    }
}
```
This is my blade file

```
@extends('layouts.app')

@section('content')

<div class="md:mx-4 relative overflow-hidden">
    <main class="h-full flex flex-col overflow-auto">
        @livewire('Livewire.TasksBoard', ['task' => $tasks])
    </main>
</div>

@endsection
```

I have tried all these possible solutions but none of them work

Attempt 1

`php artisan livewire:discover`

Attempt 2

`@livewire('Livewire.TasksBoard', ['task' => $tasks])`

Attempt 3

`@livewire('Livewire.Tasks-Board', ['task' => $tasks])`

Attempt 4

`@livewire('Livewire.TasksBoard')`

Attempt 5

`@livewire('Livewire.Tasks-Board')`

**Comment by mattkibbler** [comment, marked as answer]
Hi @h6585,

You don't need to put `Livewire` in the `@livewire` directive.

For your case you can simply use `@livewire('tasks-board', ['task' => $tasks])`.

**Reply by hussain-attari** [reply]
HI,

This didn't work either.

Although I tried as a single word before without the hyphen. Tried it again with the hyphen Tasks-Board but didn't work.

Is it a problem with capital / small case of the letters? I don't think it should be but asking it if in case.

**Reply by mattkibbler** [reply]
Yeah, you need to use all lowercase I think.

`@livewire('Tasks-Board')` won't work but `@livewire('tasks-board')` will.

You can also use tag syntax: `<livewire:tasks-board />`

**Reply by hussain-attari** [reply]
I 'think' this works. 

Can't so for sure right now as I get a different error now

```
Undefined variable: tasks (View: my-path/resources/views/tasks/index.blade.php) 

$tasks is undefined 
```

already defined it in the User Model

```
    public function tasks()
    {
        return $this->hasMany(Task::class);
    }
```

and in the status model. All the code is mentioned in this [link](https://stackoverflow.com/questions/69788866/illuminate-database-eloquent-relationnotfoundexception-call-to-undefined-relat)

**Reply by mattkibbler** [reply]
Ok, so the Livewire problem is solved? Nice!

Sounds like this next problem is that the `$tasks` variable isn't available within your blade template.

Looking at the code in the StackOverflow post you made, you are defining tasks in the index method:

```php
public function index()
{
    $tasks = auth()->user()->statuses()->with('tasks')->get();
    return view('tasks.index', compact('tasks'));
}
```
Is this where the error is occurring?

**Reply by hussain-attari** [reply]
I am not sure where the error is.

I gave the stackoverflow link as the code is already posted there.

**Reply by mattkibbler** [reply]
Where is this error happening? Your TaskController index method is setting the `$tasks` variable for the tasks.index.blade.php template, so I assume you're looking at a different page that is using the same blade template but it is not setting the `$tasks` variable?

It's difficult to say without seeing more of your code.

**Reply by hussain-attari** [reply]
What other code would be required for me to share?

Will share the layout files and the index.blade.php file in a few hours as it is on a different system.

Any other code you shall need?

Edit
As an after thought it could be that there would be an error in the index.blade.php as the data has been copied from somewhere else and pasted to create the file.

Will share the code hope you can shed some light.

**Reply by hussain-attari** [reply]
Hi,

The error is thrown at public function render ()

```
<?php

namespace App\Http\Livewire;

use Livewire\Component;

class TasksBoard extends Component
{
    public $tasks;

    public function mount()
    {
        $this->tasks = auth()->user()->statuses()->with('tasks')->get();
    }

    public function render()
    {
        return view('livewire.tasks-board', ['task' => $tasks]);
    }
}
```

**Comment by joshhanley** [comment]
@h6585 change this

```blade
@livewire('tasks-board', ['task' => $tasks])
```

To this

```blade
@livewire('tasks-board')
```

The reason is, you don't need to pass `$tasks` into the Livewire component as you are populating tasks in your `mount` method.

Hope this helps!

**Reply by hussain-attari** [reply]
Thanks I think this is the answer but can't say for sure as I am still unable to display the correct page.

The error I get is `$tasks` is undefined

**Reply by joshhanley** [reply]
@h6585 you also need to change this

```php
public function render()
    {
        return view('livewire.tasks-board', ['task' => $tasks]);
    }
 ```
 
 To this
 ```php
 public function render()
    {
        return view('livewire.tasks-board');
    }
 ```
 
 You don't need to pass `tasks` in as any public properties on the class are automatically made available to the view.
 
 Hope this helps!

**Reply by hussain-attari** [reply]
This removed that error and threw in another error.

`Invalid argument supplied for foreach() (View: my-path/resources/views/livewire/tasks-board.blade.php) `

Here is the code for the file

```
<div>
    <div class="relative p-2 flex overflow-x-auto h-full">
        <!-- Columns (Statuses) -->
        <div
        key="status.slug"
        class="mr-6 w-4/5 max-w-xs flex-1 flex-shrink-0"
        >
            @foreach ($statuses as $status)
            <div class="rounded-md shadow-md overflow-hidden">
                <div class="p-3 flex justify-between items-baseline bg-blue-800 ">
                <h4 class="font-medium text-white">
                    {{ $status->title }}
                </h4>
                <button class="py-1 px-2 text-sm text-orange-500 hover:underline">
                    Add Task
                </button>
                </div>
                <div class="p-2 flex-1 flex flex-col h-full overflow-x-hidden overflow-y-auto bg-blue-100">

                <!-- Tasks -->
                <div
                key="task.id"
                class="mb-3 p-3 h-24 flex flex-col bg-white rounded-md shadow transform hover:shadow-md cursor-pointer"
                >
                    @foreach ($status->task as $task)
                    <span class="block mb-2 text-xl text-gray-900">
                        {{ $task->title }}
                    </span>
                    <p class="text-gray-700 truncate">
                        {{ $task->description }}
                    </p>
                    @endforeach
                </div>
                <!-- ./Tasks -->

                <!-- No Tasks -->
                <div
                    x-show="!status.tasks.length"
                    class="flex-1 p-4 flex flex-col items-center justify-center"
                >
                    <span class="text-gray-600">No tasks yet!</span>
                    <button class="mt-1 text-sm text-orange-600 hover:underline">
                        Add one
                    </button>
                </div>
                <!-- ./No Tasks -->
                </div>
            </div>
            @endforeach

        </div>
        <!-- ./Columns -->

    </div>
</div>
```

The above code is originally a VueJS code which I have converted into a livewire component to suit my project needs. Probably that is why I am getting these error.

Edit:
Here is the original VueJS Script. I added the Livewire-VueJS CDN to make it work. The path opens but shows a blank page.  Also, I would prefer the code to be in pure Livewire rather than a hybrid which is why I tried converting the script to Livewire

```
<template>
  <div class="relative p-2 flex overflow-x-auto h-full">

    <!-- Columns (Statuses) -->
    <div
      v-for="status in statuses"
      :key="status.slug"
      class="mr-6 w-4/5 max-w-xs flex-1 flex-shrink-0"
    >
      <div class="rounded-md shadow-md overflow-hidden">
        <div class="p-3 flex justify-between items-baseline bg-blue-800 ">
          <h4 class="font-medium text-white">
            {{ ('status.title') }}
          </h4>
          <button class="py-1 px-2 text-sm text-orange-500 hover:underline">
            Add Task
          </button>
        </div>
        <div class="p-2 flex-1 flex flex-col h-full overflow-x-hidden overflow-y-auto bg-blue-100">
            <AddTaskForm
                v-if="newTaskForStatus === status.id"
                :status-id="status.id"
                v-on:task-added="handleTaskAdded"
                v-on:task-canceled="closeAddTaskForm"
            />
          <!-- Tasks -->
          <div
            v-for="task in status.tasks"
            :key="task.id"
            class="mb-3 p-3 h-24 flex flex-col bg-white rounded-md shadow transform hover:shadow-md cursor-pointer"
          >
            <span class="block mb-2 text-xl text-gray-900">
              {{ ('task.title') }}
            </span>
            <p class="text-gray-700 truncate">
              {{ ('task.description') }}
            </p>
          </div>
          <!-- ./Tasks -->

          <!-- No Tasks -->
          <div
            v-show="!status.tasks.length && newTaskForStatus !== status.id"
            class="flex-1 p-4 flex flex-col items-center justify-center"

            <span class="text-gray-600">No tasks yet</span>
            <button
            class="mt-1 text-sm text-orange-600 hover:underline"
            @click="openAddTaskForm(status.id)"
            >
            Add one
            </button>
            </div>
          <!-- ./No Tasks -->
        </div>
      </div>
    </div>
    <!-- ./Columns -->

  </div>
</template>

<script>
<!-- Start Added Livewire-VueJS script here ->
<script src="https://cdn.jsdelivr.net/gh/livewire/vue@v0.3.x/dist/livewire-vue.js"></script>
<!-- End Added Livewire-VueJS script here ->
export default {
  props: {
    initialData: Array
  },
  data() {
    return {
      statuses: []
    };
  },
  mounted() {
    // 'clone' the statuses so we don't alter the prop when making changes
    this.statuses = JSON.parse(JSON.stringify(this.initialData));
  }
};
</script>
```

**Reply by mattkibbler** [reply]
With this line `@foreach ($statuses as $status)` you're trying to iterate over a `$statuses` variable, but it doesn't look like you're setting that variable anywhere in your component.

Looking at the original Vue code you provided you can see that `statuses` is being set up in the `data` method and then updated in the `mounted` method. The equivalent for Livewire would be something like:

```php
<?php

namespace App\Http\Livewire;

use Livewire\Component;

class TasksBoard extends Component
{
    public $tasks;
    public $statuses;

    public function mount()
    {
        $this->tasks = auth()->user()->statuses()->with('tasks')->get();

        $this->statuses = auth()->user()->statuses();
    }

    public function render()
    {
        return view('livewire.tasks-board');
    }
}
```

**Reply by hussain-attari** [reply]
Do the statuses need to have a value?

As this error is thrown
`
Livewire component's [tasks-board] public property [statuses] must be of type: [numeric, string, array, null, or boolean]. Only protected or private properties can be set as other types because JavaScript doesn't need to access them`

**Reply by mattkibbler** [reply]
Yeah, the type of value does matter. From the [this page](https://laravel-livewire.com/docs/2.x/properties) of the Livewire docs:

> Properties can ONLY be either JavaScript-friendly data types (string, int, array, boolean), OR one of the following PHP types: Stringable, Collection, DateTime, Model, EloquentCollection.

**Reply by hussain-attari** [reply]
Is this correct?

`$this->statuses = auth()->user()->tasks()->with('statuses')->get();`

**Reply by mattkibbler** [reply]
That looks correct, as it should be returning an Eloquent Collection. Does it work for you?

**Reply by hussain-attari** [reply]
Unfortunately it doesn't. It shows a page but no data

**Reply by mattkibbler** [reply]
But you don't get any errors? That suggests that there simply aren't any results in the `$statuses` property.

Just so you know `$this->statuses = auth()->user()->tasks()->with('statuses')->get()` won't give you a list of "statuses" in the `$statuses` property, it will give you a list of "tasks" with their associates "statuses".

If you use Chrome, [this extension](https://chrome.google.com/webstore/detail/livewire-devtools/ahcmcdmhdcgbpklkdhpejphjekpmhkll?hl=en) might help. It allows you to select the Livewire components that are active on the page and inspect their data, this could help you see whether the `$statuses` property contains any data. Alternatively, you could simply do `dd($this->statuses)` in the render method and see what it contains that way.

**Reply by hussain-attari** [reply]
Yes, `it will give you a list of "tasks" with their associated "statuses"` this is exactly what I am trying to achieve.

What I am creating or rather trying to create is a Tasks board in Livewire. I still have to work out the Add a Task form part and how the status of the task is populated in the correct column.

Would appreciate it if you could point me in the right direction with how I can achieve this.

Hoping to achieve the drag and drop functionality with Alpine JS

**Reply by mattkibbler** [reply]
I'm glad your issues are now sorted.

I would recommend using Google to find existing examples of what you would like to achieve, for instance, "alpine js drag and drop" quickly led me to this example: https://codepen.io/trovster/pen/oNjGGMq

Good luck with your project!

**Reply by hussain-attari** [reply]
Thank you for all the help by you and @joshhanley 

Yes, I saw this and saw another example wherein the sort option was provided. Like this: https://codepen.io/ranjan-purbey/pen/xoEMOM

I am just trying to figure out a way to insert the data that is entered through the form to the correct column and then bind the data with the status and the user.
